commit des Blogeintrags Testautomatisierung in der Praxis (#185)
adessoAG/devblog: _posts/2020-11-17-Testautomatisierung-in-der-Praxis.md

---
layout: [post, post-xml]              
title:  "Testautomatisierung in der Praxis "        
date:   2020-11-71 15:00                    
author: axPeTo                    

> categories: [Branchen]             

Reviewer: ```suggestion
categories: [Inside adesso]             
```
Die Kategorie Branchen existiert nicht mehr, sie wurde umbenannt in Inside adesso

Reviewer 2: Branchen ist korrekt.

tags: [Testautomatisierung, Webdriver IO, Cucumber, Chai, GUI-Testautomatisierung]
---

#Testautomatisierung in der Praxis
##Erfahrungen mit GUI-Oberflächentests im Public Umfeld

> Wie bereits in unserem **letzten Blogbeitrag [LINK](https://www.adesso.de/de/news/blog/testautomatisierung-in-der-praxis-erfahrungen-mit-soapui-im-public-umfeld.jsp)** **ANGETEASERT**, handelt es sich hierbei um die Fortsetzung des gleichnamigen Blogbeitrages. Die beiden Blogbeiträge bilden eine zusammenhängende Serie, jedoch sind die Themeninhalte voneinander abgetrennt. Daher ist auch eine Einzelbetrachtung der angesprochenen Themenbereiche durchaus sinnvoll.   

Reviewer: ```suggestion
Wie bereits in unserem [letzten Blogbeitrag](https://www.adesso.de/de/news/blog/testautomatisierung-in-der-praxis-erfahrungen-mit-soapui-im-public-umfeld.jsp) angeteasert, handelt es sich hierbei um die Fortsetzung des gleichnamigen Blogbeitrages. Die beiden Blogbeiträge bilden eine zusammenhängende Serie, jedoch sind die Themeninhalte voneinander abgetrennt. Daher ist auch eine Einzelbetrachtung der angesprochenen Themenbereiche durchaus sinnvoll.   
```

> Im **ZWEITEN TEIL** des Blogs über Testautomatisierung widmen wir uns der Automation von Oberflächentests, den sogenannten GUI- bzw. UI-Tests. Wie bereits in **Teil 1** dargelegt, bietet die Testautomatisierung im Kontext der Qualitätssicherung einen deutlichen Mehrwert zur Steigerung der Softwarequalität. Trotz des hohen initialen Ressourcenaufwandes, den hohen Anforderungen an einen sensiblen Hochsicherheitsbereich und der Erstellung der technischen Infrastruktur, überwiegt die im Betrieb enorm hohe Zeitersparnis und Motivation (durch Auflösung von sich wiederholenden und monotonen Tätigkeiten), sodass sich ein Return-of-Investment (ROI) eingestellt hat. Folglich bleibt es unser Ziel, das Thema der Testautomatisierung perspektivisch und proaktiv weiter voran zu treiben und das Portfolioelement der Testautomatisierung fachverfahrensübergreifend auszubauen. 

Reviewer: Der Satz, der mit "Trotz des hohen..." anfängt ist sehr lang, unterteilt ihn bitte in mehrere Sätze. Außerdem gehört er mMn. eher ins Fazit als in die Einleitung.

Eine weitere Ausbaustufe unseres Testautomatisierungsportfolios, der wir uns in diesem Blog-Eintrag widmen, sind automatisierte Oberflächentests (fortan GUI-Tests). Aufgrund der ersten Erfolge der automatisierten Schnittstellen-Regressionstests mit Hilfe von SOAP UI Pro, wollen wir das Wissen nutzen, um diesen Ansatz bei GUI-Regressionstests weiter zu verfolgen. Folgende Fragestellungen wollen wir beleuchten: Warum wollen wir, neben automatisierten Schnittstellentests, GUI-Tests durchführen? Warum besteht die Notwendigkeit automatisierter GUI-Tests? Wie lassen sich automatisierte GUI-Tests sinnvoll einsetzen und welche Probleme tauchen dabei auf? Können wir aus den Erfahrungen der vorangegangenen Testautomatisierung und dem dort angesprochenen 4-Phasen-Modell einen Nutzen für die Oberflächentests ziehen? Lassen sich die vorliegenden Ergebnisse der Testautomatisierung von Schnittstellen mit der Testautomatisierung von Oberflächentests sinnvoll miteinander verbinden bzw. ergänzen?

###Testautomatisierung der GUI bzw. UI

> Die Anwendungen zur GUI-Automatisierung basieren primär auf Internet-/Webtechnologien und sind die Grundlage für zahlreiche Testautomatisierungstechnologien wie Webdriver IO, Python, Tosca, Selenium oder Ranorex. Diese können bei Bedarf mit anderen Tools zur (Schnittstellen-)Testautomatisierung (in unserem Fall SOAP UI Pro) kombiniert werden, indem bspw. zwischen diesen speziell angepasste Skripte ausgetauscht werden. Die Einbindung und Ausführung der Testskripte wurde entsprechend übergreifend durch SOAP gesteuert/orchestriert und bietet somit u. a. die Möglichkeit, den Testablauf an einem beliebigen Punkt der automatisierten GUI-Ausführung zu starten. Wir haben uns an dieser Stelle für das Open Source Framework Webdriver IO entschieden, welches ausschließlich den Oberflächenanteil abdeckt und ein Note.JS basiertes Testframework abbildet – somit ist er als Nachfolger von Selenium zu sehen. Doch was kann **Webdriver IO** eigentlich sonst noch? 

Reviewer: ```suggestion
Die Anwendungen zur GUI-Automatisierung basieren primär auf Internet-/Webtechnologien und sind die Grundlage für zahlreiche Testautomatisierungstechnologien wie Webdriver IO, Python, Tosca, Selenium oder Ranorex. Diese können bei Bedarf mit anderen Tools zur (Schnittstellen-)Testautomatisierung, in unserem Fall SOAP UI Pro, kombiniert werden, indem bspw. zwischen den Schnittstellen speziell angepasste Skripte ausgetauscht werden. Die Einbindung und Ausführung der Testskripte wurde entsprechend übergreifend durch SOAP gesteuert/orchestriert und bietet somit u. a. die Möglichkeit, den Testablauf an einem beliebigen Punkt der automatisierten GUI-Ausführung zu starten. Wir haben uns an dieser Stelle für das Open Source Framework Webdriver IO entschieden, welches ausschließlich den Oberflächenanteil abdeckt und ein Node.JS basiertes Testframework abbildet. Somit ist er als Nachfolger von Selenium zu sehen. Doch was kann **Webdriver IO** eigentlich sonst noch? 
```

"und ein Node.JS basiertes Testframework abbildet" -> Abbildet passt hier nicht als Verb, vielleicht ist?

Warum ist Webdriver IO erst im letzten Absatz hervorgehoben? Es sollte entweder bei der ersten Erwähnung oder garnicht hervorgehoben werden. Wenn eine Hervorhebung erfolgt, macht das bitte in Form eines Links auf die Seite von Webdriver.

> Mit Hilfe von Webdriver IO lassen sich während des Testbetriebes bspw. eingelesene Testdaten in die zu testende Weboberfläche übertragen. Dahinter verbirgt sich der Sinn, dass die Oberfläche automatisch mit Daten befüllt wird, die zum einen an das Backend gesendet und von ggf. Web-/MicroServices verarbeitet werden können, oder Feldvalidierungen getestet werden können. Zudem gilt seine Adaptierbarkeit als ein weiterer großer Vorteil, da so die Möglichkeiten geschaffen werden, andere Werkzeuge in den Testbetrieb einzugliedern. In diesem Zusammenhang sei hier einmal, für tiefer Interessierte, auf die ausführliche und sehr hilfreiche Installationsbeschreibung der Webdriver IO Projektseite hingewiesen [Webdriver IO](https://webdriver.io/docs/gettingstarted.html).

Reviewer: ```suggestion
Mit Hilfe von Webdriver IO lassen sich während des Testbetriebes bspw. eingelesene Testdaten in die zu testende Weboberfläche übertragen. Dahinter verbirgt sich der Sinn, dass die Oberfläche automatisch mit Daten befüllt wird, die zum einen an das Backend gesendet und von ggf. Web-/MicroServices verarbeitet werden können, oder Feldvalidierungen getestet werden können. Zudem gilt seine Adaptierbarkeit als ein weiterer großer Vorteil, da so die Möglichkeiten geschaffen werden, andere Werkzeuge in den Testbetrieb einzugliedern. In diesem Zusammenhang sei hier einmal, für tiefer Interessierte, auf die ausführliche und sehr hilfreiche [Installationsbeschreibung der Webdriver IO](https://webdriver.io/docs/gettingstarted.html) Projektseite hingewiesen.
```

"zum einen" im zweiten Satz erfordert ein "zum anderen". Passt diesen Satz bitte entsrechend an.

> Insofern konnten die automatisierten GUI-Tests, die in unserem Falle mit Webdriver IO entwickelt wurden, als zusätzliche Komponente unseres Regressionstestportfolios betrachtet werden. Die dafür benötigten Testdaten sind dabei bspw. entweder in einer Datenbank gespeichert oder, wie in unserem Fall, innerhalb einer organisierten und umfangreichen Excelliste hinterlegt. Diese klassische Liste enthält alle Daten und Reglements, die für die zahlreichen fachlichen und technischen Anforderungen bzw. Testszenarien relevant sind. Die Zusammenstellung der Daten und die Festlegung der Regelparameter hat mit Abstand die meisten Implementierungsaufwände mit sich gebracht. Hier liegt auch ein weiteres Problem, das mit der Kombination so vieler Testszenarien einhergeht, die nutzerfreundliche Aufbereitung/Verwaltung der hinterlegten Regressionstestszenarien. In unserem Fall sind dies gerade noch überschaubare 354 Testszenarien, jedoch kann die Anzahl (je nach Komplexitätsgrad der zu testenden Anwendungen/Systeme) exponentiell ansteigen. Um die Flut an Regeln und geforderten Testabdeckungen bedienen zu können, bieten speziell angepasste (VBA-)Makros eine wertvolle Hilfestellung. Dies haben wir bereits umfassend nutzen dürfen, mit dem Ergebnis, dass es den Endanwendern bzw. Sachbearbeitern eigenständig ermöglicht, innerhalb eines vorgegebenen Testframeworks individuelle Szenarien zu erstellen, die innerhalb der eingefassten Regelsysteme agieren. 

Reviewer: ```suggestion
Insofern konnten die automatisierten GUI-Tests, die in unserem Falle mit Webdriver IO entwickelt wurden, als zusätzliche Komponente unseres Regressionstestportfolios betrachtet werden. Die dafür benötigten Testdaten sind dabei bspw. entweder in einer Datenbank gespeichert oder, wie in unserem Fall, innerhalb einer organisierten und umfangreichen Excelliste hinterlegt. Diese klassische Liste enthält alle Daten und Reglements, die für die zahlreichen fachlichen und technischen Anforderungen bzw. Testszenarien relevant sind. Die Zusammenstellung der Daten und die Festlegung der Regelparameter hat mit Abstand die meisten Implementierungsaufwände mit sich gebracht. Die nutzerfreundliche Aufbereitung/Verwaltung der hinterlegten Regressionstestszenarien ist ein weiteres Problem, das mit der Kombination so vieler Testszenarien einhergeht. In unserem Fall sind dies gerade noch überschaubare 354 Testszenarien, jedoch kann die Anzahl an Tests je nach Komplexitätsgrad der zu testenden Anwendungen/Systeme exponentiell ansteigen. Um die Flut an Regeln und geforderten Testabdeckungen bedienen zu können, bieten speziell angepasste (VBA-)Makros eine wertvolle Hilfestellung. Dies haben wir bereits umfassend nutzen dürfen, mit dem Ergebnis, dass Endanwendende bzw. Sachbearbeitende eigenständig innerhalb eines vorgegebenen Testframeworks individuelle Szenarien erstellen können, die innerhalb der eingefassten Regelsysteme agieren. 
```

![Abbildung 1 Testdatenerstellung](/assets/images/posts/Testautomatisierung-in-der-Praxis/Testdatenerstellung.png)

> Insofern konnte hier zwar ein Teil der elementar benötigten Testaufwände verschoben werden, sodass sich die Testautomatisierer weiteren spezielleren Testszenarien widmen konnten. In diesem Zuge stieg damit auch immer weiter der bereits angesprochene Komplexitätsgrad. Abhilfe kann hier **Cucumber** schaffen. Was? Wie? Eine Gurke?

Reviewer: Das "zwar" impliziert, dass noch ein "aber" kommt, ihr müsstet den Satz umformulieren, am besten streicht ihr das "zwar".

Sind Testautomatisierer die Leute die automatisierte Tests schreiben? Bitte formuliert das entweder um oder gebt eine kurze Erklärung zu dem Begriff an. Außerdem müsstet ihr hier gendern.

Streicht bitte das "immer weiter" im zweiten Satz oder schreibt es ans Satzende.

Ihr geht hier auf den Komplexitätsgrad ein, ist das der Komplexitätsgrad der Anwendung, den ihr oben angesprochen habt? Wenn ja, in wie weit beeinflussen die Tests die Komplexität der Anwendung? Die Anwendungslogik sollte doch unabhängig von den Tests sein, oder nicht?

> Cucumber ist natürlich keine Gurke, sondern ein Behavior-Driven-Development-Werkzeug zur textuellen Spezifikation von Anforderungen. Bezogen auf unsere Projektsituation bietet es somit die Option, dem Fachbereich des Kunden eine natürlich sprachlich textuelle Testanforderungsunterstützung zur Verfügung zu stellen. Komplexe Testfälle können nach diesem Prinzip mit natürlicher Sprache präzise beschrieben werden - „Wenn Fall x eintritt dann erwarte y“. 

Reviewer: ```suggestion
Cucumber ist natürlich keine Gurke, sondern ein Behavior-Driven-Development-Werkzeug zur textuellen Spezifikation von Anforderungen. Bezogen auf unsere Projektsituation bietet es somit die Option, den Fachbereich des Kunden um eine textuelle Testanforderungsunterstützung in natürlicher Sprache zu erweitern. Komplexe Testfälle können nach diesem Prinzip in natürlicher Sprache präzise beschrieben werden - „Wenn Fall x eintritt dann erwarte y“. 
```

![Abbildung 2](/assets/images/posts/Testautomatisierung-in-der-Praxis/Testfallbeschreibung.png)

> Die so formulierten Testfälle können fortan im Hintergrund mit JavaScript ausprogrammiert werden. Mit diesem Bestreben entsteht – im besten Fall – eine Symbiose zwischen Fachbereich und der IT, die unterschiedlich formulierte Abstraktionen (Entwickler-/ Schnittstellen-/ Oberflächentests) zusammenführt und erfahrbar macht. Somit entfällt das Problem, dass auf der Entwicklungsseite Tests granular, auf einzelne Werte fokussiert geschrieben werden, die für nicht-Entwickler kryptisch erscheinen mögen und für diese Gruppe folglich nicht transparent und ohne Weiteres editierbar sind.

Reviewer: ```suggestion
Die so formulierten Testfälle können fortan im Hintergrund mit JavaScript ausprogrammiert werden. Mit diesem Bestreben entsteht – im besten Fall – eine Symbiose zwischen Fachbereich und der IT, die unterschiedlich formulierte Abstraktionen (Entwicklungs-/ Schnittstellen-/ Oberflächentests) zusammenführt und erfahrbar macht. Somit entfällt das Problem, dass auf der Entwicklungsseite Tests granular, auf einzelne Werte fokussiert geschrieben werden, die für nicht-Entwickelnde kryptisch erscheinen mögen und für diese Gruppe folglich nicht transparent und ohne Weiteres editierbar sind.
```

Ihr schreibt, dass Abstraktionen (Tests) erfahrbar gemacht werden. Was genau soll das bedeuten?

> Aus unserer Sicht erscheint es ratsam, diese Abwägung projektspezifisch zu treffen, da eine Übergabe eine enorme Einarbeitung auch abseits der projektspezifischen Charakteristika bedeutet (bspw. beim Umgang mit den verschiedenen Tools) und stetig auf eine teils nicht zu unterschätzende Wartungsleistung angewiesen ist. 

Reviewer: Ihr schreibt hier von einer Abwägung und einer Übergabe, bisher fanden keine Entscheidungen statt und von etwas, das zu übergeben sein könnte, war auch noch nicht die Rede. Dieser Satz passt vom Zusammenhang her nicht an diese Stelle, bitte entfernt ihn oder formuliert die Sätze davor so um, dass es einen inhaltlichen Zusammenhang gibt.

> Für die zukünftige Weiterentwicklung unserer Testautomatisierungsbestrebungen schwebt uns die Integration von **Chai** vor, denn dies würde es uns erlauben unsere 354 Testszenarien in Reports auswerten zu lassen. Hier sind wir allerdings derzeit noch in der Analysephase. Daher steht ein möglicher Implementierungsaufwand im Verhältnis zum Nutzen? 

Reviewer: Entweder muss der letzte Satz zu einer Frage umformuliert werden, oder das "?" muss durch einen "." ersetzt werden.

> Chai bietet die Möglichkeit, eine Auswertung zu den durchgelaufenen Tests zu erstellen. Daher können nach allen Testläufen oder selbst innerhalb einzelner Läufe entsprechend aussagekräftige Reports erstellt werden, die Basis-Informationen generieren im Hinblick auf diverse Parameter wie bspw. der Gesamtanzahl von durchgeführten Tests und derer Teilmengen von positiven und negativen Testergebnissen.

Reviewer: ```suggestion
Chai bietet die Möglichkeit, eine Auswertung zu den durchgelaufenen Tests zu erstellen. Daher können nach allen Testläufen oder selbst innerhalb einzelner Läufe entsprechend aussagekräftige Reports erstellt werden, die im Hinblick auf diverse Parameter, bspw. der Gesamtanzahl von durchgeführten Tests und der Teilmenge von positiven und negativen Testergebnissen, Basis-Informationen generieren.
```

####Unser Fazit
Nach Inbetriebnahme eines Release der von uns betreuten Fachanwendung traten unerwartet und für uns völlig überraschend Fehler in den Eingabemasken auf. Aufgrund der Kritikalität wurde das Release zurückgezogen, gefixed und ein Test aller Masken der Anwendung mit höchster Priorität angeordnet. Dies führte zu enormen manuellen Aufwänden und einer hohen Frustration, da sich monotone Tätigkeiten häuften. Nach ersten Erfolgen mit Teilautomatisierungen innerhalb der GUI, stellte sich eine erwartete Zeitersparnis anfangs jedoch nicht ein, da ständige Zertifikats- und Credentialwechsel mit begrenzter Gültigkeitsdauer auf den unterschiedlichen Testumgebungen auch für “stresserprobte” Consultants eine ambitionierte Aufgabe darstellten. Jedoch boten sich auch Überschneidungen zu vorher nicht erwarteten Bereichen, bspw. bei der Aufteilung von Testschritten, Testcases und Testsuites. Diese verglichen wir mit den bereits bekannten Unterteilungen aus dem agilen Kontext Tasks, Anforderungen und Epics, was die Akzeptanz der Umstellungen beschleunigte.

> Insgesamt erscheint (in einer konservativen Schätzung) eine **Zeitersparnis von 35%** durch den Einsatz unserer Automatisierungstätigkeiten als Durchschnittswert in unseren Projekten als realistisch: darin sind bereits der Aufwand zur intensiven Einarbeitung in SoapUI, in die anderen angesprochenen Tools und die Gespräche im Zusammenhang mit der “Überzeugungsarbeit” enthalten. Nebenbei führten die bereitgestellten Testdatenskripte zu einer wichtigen Verkürzung des Testzyklus und unterstützen agile Methoden, da einzelne zeitkritisch geforderte Testergebnisse fokussiert gewonnen und bereitgestellt werden können. Als Kunden-Feedback häuften sich die Wertschätzungen der einheitlichen und nachvollziehbaren Dokumentationen. 

Reviewer: ```suggestion
Insgesamt erscheint nach einer konservativen Schätzung eine **Zeitersparnis von 35%** durch den Einsatz unserer Automatisierungstätigkeiten als Durchschnittswert in unseren Projekten als realistisch. Darin sind bereits der Aufwand zur intensiven Einarbeitung in SoapUI, in die anderen angesprochenen Tools und die Gespräche im Zusammenhang mit der “Überzeugungsarbeit” enthalten. Nebenbei führten die bereitgestellten Testdatenskripte zu einer wichtigen Verkürzung des Testzyklus und unterstützen agile Methoden, da einzelne zeitkritisch geforderte Testergebnisse fokussiert gewonnen und bereitgestellt werden können. Als Kunden-Feedback häuften sich die Wertschätzungen der einheitlichen und nachvollziehbaren Dokumentationen. 
```

---
layout: [post, post-xml]              
title:  "Testautomatisierung in der Praxis "        
date:   2020-11-71 15:00                    
author: axPeTo                    
categories: [Inside adesso]             
tags: [Testautomatisierung, Webdriver IO, Cucumber, Chai, GUI-Testautomatisierung]
---

> 

Reviewer: ```suggestion

Als Berater für den Aufbau eines geeigneten Testumfeldes im Public Umfeld, wollen wir Euch in unserem mehrteiligen Erfahrungsbericht einladen, die verschiedenen Höhen und Tiefen in der Umsetzung zu begleiten.
In diesem Teil geht es weiter mit der Testautomatisierung, speziell im Bereich des GUI-Tests.

```

#Testautomatisierung in der Praxis
##Erfahrungen mit GUI-Oberflächentests im Public Umfeld
Wie bereits in unserem [letzten Blogbeitrag](https://www.adesso.de/de/news/blog/testautomatisierung-in-der-praxis-erfahrungen-mit-soapui-im-public-umfeld.jsp) angeteasert, handelt es sich hierbei um die Fortsetzung des gleichnamigen Blogbeitrages. 
Die beiden Blogbeiträge bilden eine zusammenhängende Serie, jedoch sind die Themeninhalte voneinander abgetrennt. 
Daher ist auch eine Einzelbetrachtung der angesprochenen Themenbereiche durchaus sinnvoll.

> Im **ZWEITEN TEIL** des Blogs über Testautomatisierung widmen wir uns der Automation von Oberflächentests, den sogenannten GUI- bzw. UI-Tests. 

Reviewer: ```suggestion
Im **Teil 2** des Blogs über Testautomatisierung widmen wir uns der Automation von Oberflächentests, den sogenannten GUI- bzw. UI-Tests. 
```

> Wie bereits in **Teil 1** dargelegt, bietet die Testautomatisierung im Kontext der Qualitätssicherung einen deutlichen Mehrwert zur Steigerung der Softwarequalität. 

Reviewer: ```suggestion
Wie bereits in **Teil 1** dargelegt, bietet die Testautomatisierung im Kontext der Qualitätssicherung großes Potential zur Steigerung der Softwarequalität. 
```

> Trotz des hohen initialen Ressourcenaufwandes, den hohen Anforderungen an einen sensiblen Hochsicherheitsbereich und der Erstellung der technischen Infrastruktur, überwiegt die im Betrieb enorm hohe Zeitersparnis und Motivation (durch Auflösung von sich wiederholenden und monotonen Tätigkeiten), sodass sich ein Return-of-Investment (ROI) eingestellt hat. 

Reviewer: ```suggestion
Trotz des hohen initialen Ressourcenaufwandes, den hohen Anforderungen an einen sensiblen Hochsicherheitsbereich und der Erstellung der technischen Infrastruktur überwiegt die im Betrieb enorm hohe Zeitersparnis und Motivation (durch Auflösung von sich wiederholenden und monotonen Tätigkeiten), sodass sich ein Return-of-Investment (ROI) eingestellt hat. 
```

Folglich bleibt es unser Ziel, das Thema der Testautomatisierung perspektivisch und proaktiv weiter voran zu treiben und das Portfolioelement der Testautomatisierung fachverfahrensübergreifend auszubauen. 
Eine weitere Ausbaustufe unseres Testautomatisierungsportfolios, der wir uns in diesem Blog-Eintrag widmen, sind automatisierte Oberflächentests (fortan GUI-Tests). 

> Aufgrund der ersten Erfolge der automatisierten Schnittstellen-Regressionstests mit Hilfe von SoapUI Pro, wollen wir das Wissen nutzen, um diesen Ansatz bei GUI-Regressionstests weiter zu verfolgen. 

Reviewer: ```suggestion
Aufgrund der ersten Erfolge der automatisierten Schnittstellen-Regressionstests mit Hilfe von SoapUI Pro, wollen wir das dort gewonnene Wissen nutzen, um diesen Ansatz bei GUI-Regressionstests weiter zu verfolgen. 
```

Folgende Fragestellungen wollen wir beleuchten: Warum wollen wir, neben automatisierten Schnittstellentests, GUI-Tests durchführen? Warum besteht die Notwendigkeit automatisierter GUI-Tests? 
Wie lassen sich automatisierte GUI-Tests sinnvoll einsetzen und welche Probleme tauchen dabei auf? 
Können wir aus den Erfahrungen der vorangegangenen Testautomatisierung und dem dort angesprochenen 4-Phasen-Modell einen Nutzen für die Oberflächentests ziehen? Lassen sich die vorliegenden Ergebnisse der Testautomatisierung von Schnittstellen mit der Testautomatisierung von Oberflächentests sinnvoll miteinander verbinden bzw. ergänzen?

###Testautomatisierung der GUI bzw. UI

> Die Anwendungen zur GUI-Automatisierung basieren primär auf Internet-/Webtechnologien und sind die Grundlage für zahlreiche Testautomatisierungstechnologien wie *Webdriver IO*, Python, Tosca, Selenium oder Ranorex. 

Reviewer: Gibt es einen bestimmten Grund, das sonst zur Hervorhebung fett gedruckt (\*\*Text**) wird und hier kursiv gedruckt (\*Text*) wird?

> Die Anwendungen zur GUI-Automatisierung basieren primär auf Internet-/Webtechnologien und sind die Grundlage für zahlreiche Testautomatisierungstechnologien wie *Webdriver IO*, Python, Tosca, Selenium oder Ranorex. 

Reviewer: ```suggestion
Die Anwendungen zur GUI-Automatisierung basieren primär auf Internet-/Webtechnologien und sind die Grundlage für zahlreiche Testautomatisierungstechnologien wie **Webdriver IO**, Python, Tosca, Selenium oder Ranorex. 
```

Diese können bei Bedarf mit anderen Tools zur (Schnittstellen-)Testautomatisierung, in unserem Fall SoapUI Pro, kombiniert werden, indem bspw. zwischen den Schnittstellen speziell angepasste Skripte ausgetauscht werden. 
Die Einbindung und Ausführung der Testskripte wurde entsprechend übergreifend durch SOAP gesteuert/orchestriert und bietet somit u. a. die Möglichkeit, den Testablauf an einem beliebigen Punkt der automatisierten GUI-Ausführung zu starten. 
Wir haben uns an dieser Stelle für das Open Source Framework Webdriver IO entschieden, welches ausschließlich den Oberflächenanteil abdeckt und ein Node.JS basiertes Testframework ist. 
Insofern ist er als Nachfolger von Selenium zu sehen. 
Doch was kann Webdriver IO eigentlich sonst noch?
Mit Hilfe von Webdriver IO lassen sich während des Testbetriebes bspw. eingelesene Testdaten in die zu testende Weboberfläche übertragen. 

> Dahinter verbirgt sich der Sinn, dass die Oberfläche automatisch mit Daten befüllt wird, die an das Backend gesendet und von ggf. Web-/MicroServices verarbeitet werden können, oder Feldvalidierungen getestet werden können. 

Reviewer: ```suggestion
Dahinter verbirgt sich der Sinn, dass die Oberfläche automatisch mit Daten befüllt wird, die an das Backend gesendet und von Web-/MicroServices verarbeitet werden können, oder dass Feldvalidierungen getestet werden kann. 
```

> Zudem gilt seine Adaptierbarkeit als ein weiterer großer Vorteil, da so die Möglichkeiten geschaffen werden, andere Werkzeuge in den Testbetrieb einzugliedern. 

Reviewer: ```suggestion
Zudem gilt seine Adaptierbarkeit als ein weiterer großer Vorteil, da so Möglichkeiten geschaffen werden, andere Werkzeuge in den Testbetrieb einzugliedern. 
```

In diesem Zusammenhang sei hier einmal, für tiefer Interessierte, auf die ausführliche und sehr hilfreiche [Installationsbeschreibung der Webdriver IO](https://webdriver.io/docs/gettingstarted.html) Projektseite hingewiesen.

Insofern konnten die automatisierten GUI-Tests, die in unserem Falle mit Webdriver IO entwickelt wurden, als zusätzliche Komponente unseres Regressionstestportfolios betrachtet werden. 
Die dafür benötigten Testdaten sind dabei bspw. entweder in einer Datenbank gespeichert oder, wie in unserem Fall, innerhalb einer organisierten und umfangreichen Excelliste hinterlegt. 
Diese klassische Liste enthält alle Daten und Reglements, die für die zahlreichen fachlichen und technischen Anforderungen bzw. Testszenarien relevant sind. 
Die Zusammenstellung der Daten und die Festlegung der Regelparameter hat mit Abstand die meisten Implementierungsaufwände mit sich gebracht. 

> Hier liegt auch ein weiteres Problem, das mit der Kombination so vieler Testszenarien einhergeht, die nutzerfreundliche Aufbereitung/Verwaltung der hinterlegten Regressionstestszenarien. 

Reviewer: ```suggestion
Hier liegt auch ein weiteres Problem, das mit der Kombination vieler Testszenarien einhergeht, die nutzerfreundliche Aufbereitung/Verwaltung der hinterlegten Regressionstestszenarien. 
```

> In unserem Fall sind dies gerade noch überschaubare 354 Testszenarien, jedoch kann die Anzahl (je nach Komplexitätsgrad der zu testenden Anwendungen/Systeme) exponentiell ansteigen. 

Reviewer: ```suggestion
In unserem Fall sind dies gerade noch überschaubare 354 Testszenarien, jedoch kann die Anzahl, je nach Komplexitätsgrad der zu testenden Anwendungen/Systeme, exponentiell ansteigen. 
```

Um die Flut an Regeln und geforderten Testabdeckungen bedienen zu können, bieten speziell angepasste (VBA-)Makros eine wertvolle Hilfestellung. 

> Dies haben wir bereits umfassend nutzen dürfen, mit dem Ergebnis, dass es den Endanwendern bzw. Sachbearbeitern eigenständig ermöglicht, innerhalb eines vorgegebenen Testframeworks individuelle Szenarien zu erstellen, die innerhalb der eingefassten Regelsysteme agieren. 

Reviewer: ```suggestion
Dies haben wir bereits umfassend nutzen dürfen, mit dem Ergebnis, dass diese Makros es den Endusern bzw. Sachbearbeitenden möglich machen, innerhalb eines vorgegebenen Testframeworks eigenständig individuelle Szenarien zu erstellen, die innerhalb der eingefassten Regelsysteme agieren. 
```

![Abbildung 1 Testdatenerstellung](/assets/images/posts/Testautomatisierung-in-der-Praxis/Testdatenerstellung.png)

> Insofern konnte hier ein Teil der elementar benötigten Testaufwände verschoben werden, sodass sich die Testautomatisierer weiteren spezielleren Testszenarien widmen konnten. 

Reviewer: ```suggestion
Insofern konnte hier ein Teil der elementar benötigten Testaufwände verschoben werden, sodass sich die Tetentwicklung weiteren spezielleren Testszenarien widmen konnten. 
```

In diesem Zuge stieg damit auch der bereits angesprochene Komplexitätsgrad im Aufbau der Tests. 
Abhilfe kann hier **Cucumber** schaffen. 
Was? Wie? Eine Gurke?
Cucumber ist natürlich keine Gurke, sondern ein Behavior-Driven-Development-Werkzeug zur textuellen Spezifikation von Anforderungen. 
Bezogen auf unsere Projektsituation bietet es somit die Option, den Fachbereich des Kunden um eine textuelle Testanforderungsunterstützung in natürlicher Sprache zu erweitern. 

> Komplexe Testfälle können nach diesem Prinzip in natürlicher Sprache präzise beschrieben werden - „Wenn Fall x eintritt dann erwarte y“. 

Reviewer: ```suggestion
Komplexe Testfälle können nach diesem Prinzip in natürlicher Sprache präzise beschrieben werden - "Wenn Fall x eintritt dann erwarte y". 
```

![Abbildung 2](/assets/images/posts/Testautomatisierung-in-der-Praxis/Testfallbeschreibung.png)

Die so formulierten Testfälle können fortan im Hintergrund mit JavaScript ausprogrammiert werden. 

> Mit diesem Bestreben entsteht – im besten Fall – eine Symbiose zwischen Fachbereich und der IT, die unterschiedlich formulierte Abstraktionen (Entwicklungs-/ Schnittstellen-/ Oberflächentests) zusammenführt und erfahrbar macht. 

Reviewer: ```suggestion
Mit diesem Bestreben entsteht – im besten Fall – eine Symbiose zwischen dem Fachbereich und der IT, die unterschiedlich formulierte Abstraktionen (Entwicklungs-/ Schnittstellen-/ Oberflächentests) zusammenführt und erfahrbar macht. 
```

> Somit entfällt das Problem, dass auf der Entwicklungsseite Tests granular, auf einzelne Werte fokussiert geschrieben werden, die für nicht-Entwickelnde kryptisch erscheinen mögen und für diese Gruppe folglich nicht transparent und ohne Weiteres editierbar sind.

Reviewer: ```suggestion
Somit entfällt das Problem, dass auf der Entwicklungsseite Tests granular, auf einzelne Werte fokussiert geschrieben werden, die für nicht-Entwickelnde kryptisch erscheinen mögen und für diese Gruppe folglich nicht transparent und somit nicht ohne Weiteres editierbar sind.
```

> Aus unserer Sicht erscheint es ratsam, diese Abwägung projektspezifisch zu treffen, da eine mögliche Übergabe des Projektes eine enorme Einarbeitung auch abseits der projektspezifischen Charakteristika bedeutet (bspw. beim Umgang mit den verschiedenen Tools) und stetig auf eine teils nicht zu unterschätzende Wartungsleistung angewiesen ist.

Reviewer: ```suggestion
Aus unserer Sicht erscheint es ratsam, die Entscheidung, wie tiefgehend GUI-Testautomatisierung erfolgen soll, projektspezifisch zu treffen, da eine mögliche Übergabe des Projektes eine enorme Einarbeitung auch abseits der projektspezifischen Charakteristika bedeutet (bspw. beim Umgang mit den verschiedenen Tools) und die Tools auf eine teils nicht zu unterschätzende, stetige Wartungsleistung angewiesen sind.
```

> Für die zukünftige Weiterentwicklung unserer Testautomatisierungsbestrebungen schwebt uns die Integration von **Chai** vor, denn dies würde es uns erlauben unsere 354 Testszenarien in Reports auswerten zu lassen. 

Reviewer: ```suggestion
Für die zukünftige Weiterentwicklung unserer Testautomatisierungsbestrebungen schwebt uns die Integration von **Chai** vor, denn dies würde es uns erlauben, unsere 354 Testszenarien in Reports auswerten zu lassen. 
```

Hier sind wir allerdings derzeit noch in der Analysephase. 

> Daher kommt hier die Frage auf, ob ein möglicher Implementierungsaufwand im Verhältnis zum Nutzen steht?

Reviewer: ```suggestion
Daher kommt hier die Frage auf, ob ein möglicher Implementierungsaufwand im Verhältnis zum Nutzen steht.
```

Chai bietet die Möglichkeit, eine Auswertung zu den durchgelaufenen Tests zu erstellen. 

> Daher können nach allen Testläufen oder selbst innerhalb einzelner Läufe entsprechend aussagekräftige Reports erstellt werden, die im Hinblick auf diverse Parameter, bspw. der Gesamtanzahl von durchgeführten Tests und der Teilmenge von positiven und negativen Testergebnissen, Basis-Informationen generieren.

Reviewer: ```suggestion
Es können also nach allen Testläufen oder selbst innerhalb einzelner Läufe entsprechend aussagekräftige Reports erstellt werden, die im Hinblick auf diverse Parameter, bspw. der Gesamtanzahl von durchgeführten Tests und der Teilmenge von positiven und negativen Testergebnissen, Basis-Informationen generieren.
```

Dies würde entsprechend einen weiteren manuellen Aufbereitungsschritt obsolet machen. 
Dazu mehr in einem möglicherweise kommenden **Teil 3** unserer Serie zur Testautomatisierung im Public Umfeld.

##Unser Fazit
Nach Inbetriebnahme eines Release der von uns betreuten Fachanwendung traten unerwartet und für uns völlig überraschend Fehler in den Eingabemasken auf. 
Aufgrund der Kritikalität wurde das Release zurückgezogen, gefixed und ein Test aller Masken der Anwendung mit höchster Priorität angeordnet. 

> Dies führte zu enormen manuellen Aufwänden und einer hohen Frustration, da sich monotone Tätigkeiten häuften. 

Reviewer: ```suggestion
Dies verursachte enormen manuellen Aufwand und eine hohen Frustration, da sich monotone Tätigkeiten häuften. 
```

> Nach ersten Erfolgen mit Teilautomatisierungen innerhalb der GUI, stellte sich eine erwartete Zeitersparnis anfangs jedoch nicht ein, da ständige Zertifikats- und Credentialwechsel mit begrenzter Gültigkeitsdauer auf den unterschiedlichen Testumgebungen auch für “stresserprobte” Consultants eine ambitionierte Aufgabe darstellten. 

Reviewer: ```suggestion
Nach ersten Erfolgen mit Teilautomatisierungen innerhalb der GUI stellte sich eine erwartete Zeitersparnis anfangs jedoch nicht ein, da ständige Zertifikats- und Credentialswechsel mit begrenzter Gültigkeitsdauer auf den unterschiedlichen Testumgebungen auch für "stresserprobte" Consultants eine ambitionierte Aufgabe darstellten. 
```

Jedoch boten sich auch Überschneidungen zu vorher nicht erwarteten Bereichen, bspw. bei der Aufteilung von Testschritten, Testcases und Testsuites. 
Diese verglichen wir mit den bereits bekannten Unterteilungen aus dem agilen Kontext Tasks, Anforderungen und Epics, was die Akzeptanz der Umstellungen beschleunigte.
Insgesamt erscheint nach einer konservativen Schätzung eine **Zeitersparnis von 35%** durch den Einsatz unserer Automatisierungstätigkeiten als Durchschnittswert in unseren Projekten als realistisch. 

> Darin sind bereits der Aufwand zur intensiven Einarbeitung in SoapUI, in die anderen angesprochenen Tools und die Gespräche im Zusammenhang mit der “Überzeugungsarbeit” enthalten. 

Reviewer: ```suggestion
Darin sind bereits der Aufwand zur intensiven Einarbeitung in SoapUI, in die anderen angesprochenen Tools und die Gespräche im Zusammenhang mit der "Überzeugungsarbeit" enthalten. 
```

Nebenbei führten die bereitgestellten Testdatenskripte zu einer wichtigen Verkürzung des Testzyklus und unterstützen agile Methoden, da einzelne zeitkritisch geforderte Testergebnisse fokussiert gewonnen und bereitgestellt werden können. 

> Als Kunden-Feedback häuften sich die Wertschätzungen der einheitlichen und nachvollziehbaren Dokumentationen. 

Reviewer: ```suggestion
Als Kunden-Feedback häufte sich die Wertschätzungen der einheitlichen und nachvollziehbaren Dokumentationen. 
```

---
layout: [post, post-xml]              
title:  "Testautomatisierung in der Praxis "        

> date:   2020-11-71 15:00                    

Reviewer: ```suggestion
date:   2020-11-30 15:00                    
```

author: axPeTo                    
categories: [Inside adesso]             
tags: [Testautomatisierung, Webdriver IO, Cucumber, Chai, GUI-Testautomatisierung]
---

Als Berater für den Aufbau eines geeigneten Testumfeldes im Public Umfeld, wollen wir Euch in unserem mehrteiligen Erfahrungsbericht einladen, die verschiedenen Höhen und Tiefen in der Umsetzung zu begleiten.
In diesem Teil geht es weiter mit der Testautomatisierung, speziell im Bereich des GUI-Tests.

> #Testautomatisierung in der Praxis

Reviewer: ```suggestion
# Testautomatisierung in der Praxis
```

> ##Erfahrungen mit GUI-Oberflächentests im Public Umfeld

Reviewer: ```suggestion
## Erfahrungen mit GUI-Oberflächentests im Public Umfeld
```

> # Testautomatisierung in der Praxis
> ## Erfahrungen mit GUI-Oberflächentests im Public Umfeld

Reviewer: ```suggestion
# Erfahrungen mit GUI-Oberflächentests im Public Umfeld
```

Eine weitere Ausbaustufe unseres Testautomatisierungsportfolios, der wir uns in diesem Blog-Eintrag widmen, sind automatisierte Oberflächentests (fortan GUI-Tests). 
Aufgrund der ersten Erfolge der automatisierten Schnittstellen-Regressionstests mit Hilfe von SoapUI Pro, wollen wir das dort gewonnene Wissen nutzen, um diesen Ansatz bei GUI-Regressionstests weiter zu verfolgen. 
Folgende Fragestellungen wollen wir beleuchten: Warum wollen wir, neben automatisierten Schnittstellentests, GUI-Tests durchführen? Warum besteht die Notwendigkeit automatisierter GUI-Tests? 
Wie lassen sich automatisierte GUI-Tests sinnvoll einsetzen und welche Probleme tauchen dabei auf? 
Können wir aus den Erfahrungen der vorangegangenen Testautomatisierung und dem dort angesprochenen 4-Phasen-Modell einen Nutzen für die Oberflächentests ziehen? Lassen sich die vorliegenden Ergebnisse der Testautomatisierung von Schnittstellen mit der Testautomatisierung von Oberflächentests sinnvoll miteinander verbinden bzw. ergänzen?

> ###Testautomatisierung der GUI bzw. UI

Reviewer: ```suggestion
### Testautomatisierung der GUI bzw. UI
```

> ### Testautomatisierung der GUI bzw. UI

Reviewer: ```suggestion
## Testautomatisierung der GUI bzw. UI
```

Für die zukünftige Weiterentwicklung unserer Testautomatisierungsbestrebungen schwebt uns die Integration von **Chai** vor, denn dies würde es uns erlauben, unsere 354 Testszenarien in Reports auswerten zu lassen. 
Hier sind wir allerdings derzeit noch in der Analysephase. 
Daher kommt hier die Frage auf, ob ein möglicher Implementierungsaufwand im Verhältnis zum Nutzen steht.
Chai bietet die Möglichkeit, eine Auswertung zu den durchgelaufenen Tests zu erstellen. 
Es können also nach allen Testläufen oder selbst innerhalb einzelner Läufe entsprechend aussagekräftige Reports erstellt werden, die im Hinblick auf diverse Parameter, bspw. der Gesamtanzahl von durchgeführten Tests und der Teilmenge von positiven und negativen Testergebnissen, Basis-Informationen generieren.
Dies würde entsprechend einen weiteren manuellen Aufbereitungsschritt obsolet machen. 
Dazu mehr in einem möglicherweise kommenden **Teil 3** unserer Serie zur Testautomatisierung im Public Umfeld.

> ##Unser Fazit

Reviewer: ```suggestion
## Unser Fazit
```

> ## Unser Fazit

Reviewer: ```suggestion
# Unser Fazit
```
